File: Sources/Network/_NetworkHelper.m
//
//  Example
//  man
//
//  Created by man 11/11/2018.
//  Copyright © 2020 man. All rights reserved.
//

#import "_NetworkHelper.h"
#import "_CustomHTTPProtocol.h"
#import "NSObject+CocoaDebug.h"

@interface _NetworkHelper()

@end

@implementation _NetworkHelper

+ (instancetype)shared
{
    static id sharedInstance = nil;
    
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        sharedInstance = [[self alloc] init];
    });
    
    return sharedInstance;
}

//default value for @property
- (id)init {
    if (self = [super init])  {
        self.mainColor = [UIColor colorFromHexString:@"#42d459"];
        self.isNetworkEnable = YES;
        self.maxResponseSize = 2 * 1024 * 1024;    // 2 MB (was 10 MB - reduced to lower memory pressure)
        self.maxRequestBodySize = 512 * 1024;       // 512 KB
    }
    return self;
}

- (void)enable {
    if (self.isNetworkEnable) {
        return;
    }
    self.isNetworkEnable = YES;
    [_CustomHTTPProtocol start];
}

- (void)disable {
    if (!self.isNetworkEnable) {
        return;
    }
    self.isNetworkEnable = NO;
    [_CustomHTTPProtocol stop];
}

@end

File: Sources/Network/NetworkHelper.swift
//
//  Example
//  man
//
//  Created by man 11/11/2018.
//  Copyright © 2020 man. All rights reserved.
//

import Foundation
import UIKit

@objc class _NetworkHelper: NSObject {

    /// color for objc
    @objc var mainColor: UIColor?

    /// Set domain names not to be crawled, ignore case, and crawl all by default
    @objc var ignoredURLs: [String]?

    /// Set only the domain name to be crawled, ignore case, and crawl all by default
    @objc var onlyURLs: [String]?

    /// Set the log prefix not to be crawled, ignore case, and crawl all by default
    @objc var ignoredPrefixLogs: [String]?

    /// Set the log prefix to be crawled, ignore case, and crawl all by default
    @objc var onlyPrefixLogs: [String]?

    /// protobuf
    @objc var protobufTransferMap: [String: [String]]?

    /// Maximum response body size (bytes) to capture. Default 10 MB.
    @objc var maxResponseSize: UInt = 0

    /// Maximum request body size (bytes) to capture from streams. Default 512 KB.
    @objc var maxRequestBodySize: UInt = 0

    @objc var isNetworkEnable: Bool = false

    @objc static let shared = _NetworkHelper()

    private override init() {
        super.init()
        mainColor = "#42d459".hexColor
        // In ObjC, +load ran swizzleSessionConfiguration() before main(), so
        // isNetworkEnable=YES was safe — all sessions already had the protocol.
        // In Swift there is no +load, so we must leave isNetworkEnable=false here
        // so that enable() actually calls _CustomHTTPProtocol.start() (registerClass),
        // which covers URLSession.shared and sessions created before the swizzle.
        isNetworkEnable = false
        maxResponseSize = 10 * 1024 * 1024      // 10 MB (data is stored on disk, not in memory)
        maxRequestBodySize = 512 * 1024          // 512 KB
    }

    @objc func enable() {
        if isNetworkEnable {
            return
        }
        isNetworkEnable = true
        _CustomHTTPProtocol.start()
    }

    @objc func disable() {
        if !isNetworkEnable {
            return
        }
        isNetworkEnable = false
        _CustomHTTPProtocol.stop()
    }
}
